From ce4bd9a3e461976bc3e07484608ce279c8dad114 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 08 Oct 2020 11:14:43 +0300
Subject: [PATCH] [SW-22078] preboot: goya2: set arcs in running state

As per new requirement of loading zephyr images on ARC from BTL based
boot flow, ARC should be put to running state as part of start
image function in preboot after fit images of zephyr are copied in ARC's
ICCM and DCCM. This patch add support to put ARC0 and ARC1 in running
state which will start zephyr OS (PID and MGMT) on ARC0 and ARC1.

Change-Id: Iaed23371e504c7bf836795f93cd6d31a8c667f30
---

diff --git a/goya2/common.c b/goya2/common.c
index e604ab6..7456547 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -278,6 +278,10 @@
 		break;
 	}
 
+	/* Set ARC0 and ARC1 in running state */
+	RUN_ARC(0);
+	RUN_ARC(1);
+
 	/*
 	 * After image_entry, control is not expected to return
 	 * here and continue.
diff --git a/goya2/common.h b/goya2/common.h
index 0f82de8..e40f34c 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -34,3 +34,17 @@
 		regmap_writel(0x0, mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);	 \
 		regmap_writel(0x1, mmPSOC_ARC##id##_CFG_ARC_RST);	 \
 	} while (0)
+/*
+ * Set ARC in running state
+ * 1. Set run bit in RUN_HALT_REQ register
+ * 2. Wait for ACK from core in RUN_HALT_ACK register
+ * 3. Clear run bit in RUN_HALT_REQ register
+ */
+#define RUN_ARC(id)							       \
+	do {								       \
+		regmap_writel(BIT(0), mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);      \
+		while (!(regmap_readl(mmPSOC_ARC##id##_CFG_RUN_HALT_ACK) &     \
+				PSOC_ARC##id##_CFG_RUN_HALT_ACK_RUN_ACK_MASK)) \
+			;						       \
+		regmap_writel(0x0, mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);	       \
+	} while (0)
